modules/apps/notification/notification-service/src/main/java/com/liferay/notification/internal/upgrade/v3_10_4/DeleteStaleNotificationUpgradeProcess.java

```diff
@@ -0,0 +1,244 @@
+		_deleteNotificationTemplate();
+	}
+
+	private void _deleteNotificationQueueEntry() throws Exception {
+		try (PreparedStatement deletePreparedStatement1 =
+				 _getDeletePreparedStatement("NQueueEntryAttachment",
+					 "notificationQueueEntryId");
+			 PreparedStatement deletePreparedStatement2 =
+				 _getDeletePreparedStatement("NotificationQueueEntry",
+						 "notificationQueueEntryId");
+			 PreparedStatement selectPreparedStatement1 =
+				 _getSelectPreparedStatement();
+			 PreparedStatement selectPreparedStatement2 =
+				 _getSelectPreparedStatement("NotificationQueueEntry");
+
+			 ResultSet resultSet1 = selectPreparedStatement2.executeQuery()) {
+
+			while (resultSet1.next()) {
+				long notificationQueueEntryId = resultSet1.getLong(
+					"notificationQueueEntryId");
+				long companyId = resultSet1.getLong("companyId");
+
+				_deleteResourcePermissions(
+					NotificationQueueEntry.class.getName(),
+					notificationQueueEntryId, companyId);
+
+				deletePreparedStatement1.setLong(1, notificationQueueEntryId);
+				deletePreparedStatement2.setLong(1, notificationQueueEntryId);
+
+				deletePreparedStatement1.addBatch();
+				deletePreparedStatement2.addBatch();
+
+				selectPreparedStatement1.setLong(
+					1,
+					PortalUtil.getClassNameId(
+						NotificationQueueEntry.class.getName()));
+				selectPreparedStatement1.setLong(2, notificationQueueEntryId);
+
+				ResultSet resultSet2 = selectPreparedStatement1.executeQuery();
+
+				while (resultSet2.next()) {
+					_deleteNotificationRecipient(resultSet2.getLong(1));
+				}
+
+				Repository repository = _getRepository(companyId);
+
+				if (repository != null) {
+					try {
+						Folder folder = _portletFileRepository.getPortletFolder(
+							repository.getRepositoryId(),
+							DLFolderConstants.DEFAULT_PARENT_FOLDER_ID,
+							String.valueOf(notificationQueueEntryId));
+
+						_portletFileRepository.deletePortletFolder(
+							folder.getFolderId());
+					}
+					catch (PortalException portalException) {
+						if (_log.isDebugEnabled()) {
+							_log.debug(portalException);
+						}
+					}
+				}
+			}
+
+			deletePreparedStatement1.executeBatch();
+			deletePreparedStatement2.executeBatch();
+		}
+	}
+
+	private void _deleteNotificationRecipient(long notificationRecipientId)
+		throws Exception {
+
+		runSQL(
+			StringBundler.concat(
+				"delete from NotificationRecipient where ",
+				"notificationRecipientId = ", notificationRecipientId));
+		runSQL(
+			StringBundler.concat(
+				"delete from NotificationRecipientSetting where ",
+				"notificationRecipientId = ", notificationRecipientId));
+	}
+
+	private void _deleteNotificationTemplate() throws Exception {
+		try (PreparedStatement deletePreparedStatement1 =
+				 _getDeletePreparedStatement("NotificationTemplate",
+					 "notificationTemplateId");
+			 PreparedStatement deletePreparedStatement2 =
+				 _getDeletePreparedStatement("NTemplateAttachment",
+					 "notificationTemplateId");
```

Comment:
Can you sort it, please?
**N**TemplateAttachment > **No**tification**T**emplate
`N` should come first, then `No`.
Right?